### ISSUE TITLE
Does ModelSpeedup support model like ResNet or DenseNet?

### ISSUE THREAD

**Issue by Junwen-Huang** [microsoft/nni#2130]
**Short summary about the issue/question**: The model in examples of model compression is LeNet or VGG. And it seems that ResNet is not supported yet, because 'aten::add' or concatenate is not supported. Is there any plan to support models with shortcut connection?

**Brief what process you are following**: 

**How to reproduce it**: 

**nni Environment**:
- nni version: 1.4
- nni mode(local|pai|remote): local
- OS:Windows 10
- python version: 3.7
- is conda or virtualenv used?: conda
- is running in docker?: no

**need to update document(yes/no)**:

**Anything else we need to know**:

**Comment by Junwen-Huang** [comment]
Plus, I implement a network for MNIST classification, cascaded with conv2d and pooling with pytorch 1.3.1 and torchvison 0.4.2. I find that if BatchNorm2d is used, the model after speed up does not output the same as the masked one. In the case observed, accuracy drops down obviously.
I put the code on Google drive: [https://drive.google.com/open?id=1DXSY3_y_jn8HbYlZKKqiz2TsHizLG7yF](url).

**Comment by QuanluZhang** [comment]
@Junwen-Huang thanks for reporting this issue. The output mismatch is induced by BatchNorm2d, precisely, it is induced by non-zero `beta` in [BatchNorm2d](https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d). `beta` is non-zero because you loaded checkpoint of the model.

Take a simple example: conv1 -> batchnorm2d -> conv2. Let's say 90% out channels of conv1 are pruned, and 90% out channels of conv2 are pruned. For conv2, ModelSpeedup will prune both 0th (out channel) and 1st (in channel) dimension of its weight. However, if `beta` of batchnorm2d is not zero, pruning 1st dimension makes ModelSpeedup produce a different output from that of the masked one. You can check that if you do not load batchnorm2d's weight, the output would be the same.

1st dimension of conv2 is better to be pruned, because first it leads to smaller model, second, there is no information delivered by the pruned in channels of conv2 (i.e., the pruned out channels of conv1). To improve the model's performance after ModelSpeedup, you can retrain bn or the whole model.

**Comment by QuanluZhang** [comment]
@Junwen-Huang for your first question, i.e., supporting more operations. We encourage users to contribute more operation support (i.e., shape inference and replacement of operations). In future release, we will make adding new operation support much easier.

**Comment by Junwen-Huang** [comment]
> @Junwen-Huang thanks for reporting this issue. The output mismatch is induced by BatchNorm2d, precisely, it is induced by non-zero `beta` in [BatchNorm2d](https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d). `beta` is non-zero because you loaded checkpoint of the model.
> 
> Take a simple example: conv1 -> batchnorm2d -> conv2. Let's say 90% out channels of conv1 are pruned, and 90% out channels of conv2 are pruned. For conv2, ModelSpeedup will prune both 0th (out channel) and 1st (in channel) dimension of its weight. However, if `beta` of batchnorm2d is not zero, pruning 1st dimension makes ModelSpeedup produce a different output from that of the masked one. You can check that if you do not load batchnorm2d's weight, the output would be the same.
> 
> 1st dimension of conv2 is better to be pruned, because first it leads to smaller model, second, there is no information delivered by the pruned in channels of conv2 (i.e., the pruned out channels of conv1). To improve the model's performance after ModelSpeedup, you can retrain bn or the whole model.

Same example: conv1 -> batchnorm2d -> conv2. I find that ModelSpeedup will prune 90% out channel of conv1, 90% mean, variance, gamma and beta of batchnorm2d, 90% out channel and in channel of conv2. All those parameters pruned is indexed by the masked. Mathematically, whether beta is non-zero should not produce different output, because all parameters (conv, bn) of rest channels are the same as the masked one.

**Comment by QuanluZhang** [comment]
yes, this is how ModelSpeedup prunes this model. On the other hand, let's look at how this model is masked by masks. Only 90% out channels of conv1 and conv2 are masked respectively. Thus, 90% out channels of conv1 are zero, however, none of them are zero after batchnorm2d because `beta` is not zero. As in channels are not masked in conv2, those non-zero channels in batchnorm2d's output take effect.

you can set `beta` to 0 to check whether the output is the same.

**Comment by Junwen-Huang** [comment]
> yes, this is how ModelSpeedup prunes this model. On the other hand, let's look at how this model is masked by masks. Only 90% out channels of conv1 and conv2 are masked respectively. Thus, 90% out channels of conv1 are zero, however, none of them are zero after batchnorm2d because `beta` is not zero. As in channels are not masked in conv2, those non-zero channels in batchnorm2d's output take effect.
> 
> you can set `beta` to 0 to check whether the output is the same.

Understand. Thank you so much. Maybe it is a good idea to mask the BN in the Pruner as well in future. 
(￣▽￣)~*

**Comment by scarlett2018** [comment]
Closing the issue given the question is answered, thanks @Junwen-Huang and @QuanluZhang

**Comment by chikubee** [comment]
Hi, I encountered a similar issue. Though this is a general question, why did the model size not reduce on pruning. Is the replacement really not happening and only the tensors are set to 0.?
TIA

**Comment by QuanluZhang** [comment]
@chikubee , yes, most model pruning algorithms are based on simulation, they use mask matrix to simulate the effect of pruning. There is no real model speedup. NNI ModelSpeedup does it for them.

**Comment by ashutoshmishra1014** [comment]
@QuanluZhang I am having a similar issue. But my question is if the model speed up is already executed shouldn't the beta issue with the batchnorm layers following a pruned conv layer should already get resolved? My belief is that it should but still I am getting a drop in accuracy between masked inference and speeduped inference.

**Comment by QuanluZhang** [comment]
@ashutoshmishra1014 , currently you still need to fine-tune bn or the whole model to in order to improve model accuracy. 

Speeding up the model which has bn while keeping its accuracy will be supported in release v2.5.

In v2.5 NNI should provide much more powerful model speedup capability.

**Comment by ashutoshmishra1014** [comment]
@QuanluZhang Thanks a lot for sharing the update.
